plugins/renepay/mcf.c

```diff
@@ -0,0 +1,1507 @@
+#include "config.h"
+#include <assert.h>
+#include <ccan/list/list.h>
+#include <ccan/lqueue/lqueue.h>
+#include <ccan/tal/tal.h>
+#include <common/type_to_string.h>
+#include <math.h>
+#include <plugins/renepay/debug.h>
+#include <plugins/renepay/dijkstra.h>
+#include <plugins/renepay/flow.h>
+#include <plugins/renepay/mcf.h>
+#include <stdint.h>
+
+/* # Optimal payments
+ *
+ * In this module we reduce the routing optimization problem to a linear
+ * cost optimization problem and find a solution using MCF algorithms.
+ * The optimization of the routing itself doesn't need a precise numerical
+ * solution, since we can be happy near optimal results; e.g. paying 100 msat or
+ * 101 msat for fees doesn't make any difference if we wish to deliver 1M sats.
+ * On the other hand, we are now also considering Pickhard's
+ * [1] model to improve payment reliability,
+ * hence our optimization moves to a 2D space: either we like to maximize the
+ * probability of success of a payment or minimize the routing fees, or
+ * alternatively we construct a function of the two that gives a good compromise.
+ *
+ * Therefore from now own, the definition of optimal is a matter of choice.
+ * To simplify the API of this module, we think the best way to state the
+ * problem is:
+ *
+ * 	Find a routing solution that pays the least of fees while keeping
+ * 	the probability of success above a certain value `min_probability`.
```

Comment:
META COMMENT: `-log(Probability)` can be defined as an uncertainty cost and routing fees can be considered as a routing cost. In that case we would wish to minimize the total cost which consists of the combination of both routing and uncertainty cost. I think that is a more general and less confusing way to formulate the problem.